openedx/core/lib/gating/services.py

```diff
@@ -0,0 +1,47 @@
+"""
+A wrapper class around requested methods exposed in api.py
+"""
+import types
+
+from . import api as gating_api
+
+
+class GatingService(object):  # pylint: disable=too-few-public-methods
+    """
+    An xBlock service for xBlocks to talk to the Gating api.
+
+    NOTE: This is a Singleton class. We should only have one instance of it!
+    """
+
+    _instance = None
+
+    REQUESTED_FUNCTIONS = [
+        'get_gating_milestone_meta_info',
+        'get_gating_milestone',
+        'is_prereq_met'
+    ]
+
+    def __new__(cls, *args, **kwargs):
+        """
+        This is the class factory to make sure this is a Singleton
+        """
+        if not cls._instance:
+            cls._instance = super(GatingService, cls).__new__(cls, *args, **kwargs)
+        return cls._instance
+
+    def __init__(self):
+        """
+        Class initializer, which just inspects the libraries and exposes the same functions
+        listed in REQUESTED_FUNCTIONS
+        """
+        self._bind_to_requested_functions()
+
+    def _bind_to_requested_functions(self):
+        """
+        bind module functions. Since we use underscores to mean private methods, let's exclude those.
+        """
+        for attr_name in self.REQUESTED_FUNCTIONS:
+            attr = getattr(gating_api, attr_name, None)
+            if isinstance(attr, types.FunctionType) and not attr_name.startswith('_'):
+                if not hasattr(self, attr_name):
+                    setattr(self, attr_name, attr)
```

Comment:
Given that there are only three functions to proxy here, perhaps this class could be dumber and explicitly call out each method call? So:

```python
def is_prereq_met(self, content_id, user_id, recalc_on_unmet=False):
   return api.is_prereq_met(content_id, user_id, recalc_on_unmet)
```

Or if you're worried about repeating things, you can do:

```python
def is_prereq_met(self, *args, **kwargs):
    """See api.is_prereq_met for usage info."""
    return api.is_prereq_met(*args, **kwargs)
```

Given that you can remove the `__init__` and `_bind_to_requested_functions`, it's not really much longer. And it's less magical for people who aren't as familiar with Python.

Reply:
agree, will fix